### PR TITLE
Added backpressure to outputnsq

### PR DESCRIPTION
**Pull request by helgeolav**
I have added backpressure to outputnsq. This change required some changes to simplequeue as I also needed to handle events that was sent through an codec.

### CODE REVIEW

File: config/queue/simplequeue.go [hunk 1]
```diff
@@ -18,29 +18,30 @@ type simpleQueue struct {
 	ctx        context.Context
 	output     QueueReceiver // the output
 	control    config.Control
-	isInPause  uint32                 // set to either StatusDelivering or StatusPaused
-	queue      chan logevent.LogEvent // channel to send events to the internal queue
-	retryqueue list.List              // list of queued messages; the list is not multithreading safe, and must only be accessed from backgroundtask()
+	isInPause  uint32           // set to either StatusDelivering or StatusPaused
+	queue      chan interface{} // channel to send events to the internal queue
+	codecCh    chan []byte      // channel to push a coded message onto
+	retryqueue list.List        // list of queued messages; the list is not multithreading safe, and must only be accessed from backgroundtask()
```
Comment (tsaikd) [inline config/queue/simplequeue.go:24]: retryqueue can be defined in backgroundtask() scope

File: config/queue/simplequeue.go [hunk 2]
```diff
@@ -108,43 +110,69 @@ func (t *simpleQueue) backgroundtask() {
...
-						event := e.Value.(logevent.LogEvent)
-						myList = append(myList, &event)
+						event := e.Value
+						myList = append(myList, event)
 						t.retryqueue.Remove(e)
 					}
 					// Now we have to send them all out
 					go func() {
 						for x := range myList {
 							ctx, cancel := context.WithTimeout(context.Background(), dur)
```
Comment (tsaikd) [inline config/queue/simplequeue.go:161]: the cancel function is not used on all paths

File: config/queue/simplequeue_codec_test.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+			id := atomic.AddUint32(&c.numReceived, 1)
+			var failed bool
+			for x := range c.FailMsgId {
+				if c.FailMsgId[x] == id {
+					failed = true
+					break
+				}
+			}
+			if failed {
+				c.queue.Queue(c.ctx, msg)
```
Comment (tsaikd) [inline config/queue/simplequeue_codec_test.go:41]: Error return value of `c.queue.Queue` is not checked

**Comment by helgeolav** [comment]
Hi

I have looked through your comments and changed the code based on your comments.

**Pull request merged**